src/discordas.ts

```diff
@@ -78,53 +102,42 @@ async function run(port: number, fileConfig: DiscordBridgeConfig) {
+                    return await callbacks.onAliasQueried!(alias, roomId);
                 } catch (err) { log.error("Exception thrown while handling \"onAliasQueried\" event", err); }
             },
-            onAliasQuery: async (alias: string, aliasLocalpart: string) => {
+            onAliasQuery: async (alias: string, aliasLocalpart: string): Promise<IProvisionedRoom|undefined> => {
                 try {
-                    return await callbacks.onAliasQuery(alias, aliasLocalpart);
+                    return await callbacks.onAliasQuery!(alias, aliasLocalpart);
                 } catch (err) { log.error("Exception thrown while handling \"onAliasQuery\" event", err); }
             },
-            onEvent: async (request) => {
-                const data = request.getData();
+            onEvent: async (request: Request, _: BridgeContext): Promise<void> => {
                 try {
-                    MetricPeg.get.registerRequest(data.event_id);
-                    // Build our own context.
-                    if (!store.roomStore) {
-                        log.warn("Discord store not ready yet, dropping message");
-                        MetricPeg.get.requestOutcome(data.event_id, false, "dropped");
-                        return;
-                    }
-                    const roomId = data.room_id;
-
-                    const context: BridgeContext = {
-                        rooms: {},
-                    };
-
-                    if (roomId) {
-                        const entries  = await store.roomStore.getEntriesByMatrixId(roomId);
-                        context.rooms = entries[0] || {};
-                    }
-
-                    await request.outcomeFrom(callbacks.onEvent(request, context));
-                    MetricPeg.get.requestOutcome(data.event_id, false, "success");
+                    const event = request.getData() as IMatrixEvent;
+                    const roomId = event.room_id;
+
+                    MetricPeg.get.registerRequest(event.event_id);
+
+                    const context = await buildOwnContext(roomId, store);
+                    const callbackResult = callbacks.onEvent(request, context);
+                    request.outcomeFrom(callbackResult);
                 } catch (err) {
-                    MetricPeg.get.requestOutcome(data.event_id, false, "fail");
-                    log.error("Exception thrown while handling \"onEvent\" event", err);
-                    await request.outcomeFrom(Promise.reject("Failed to handle"));
+                    logOnEventError(err);
+                    request.reject(err);
+                } finally {
+                    recordRequestOutcome(request);
                 }
             },
-            onLog: (line, isError) => {
-                log.verbose("matrix-appservice-bridge", line);
+            onLog: (text: string, isError: boolean): void => {
+                const logText = (isError ? log.error : log.verbose).bind(log);
```

Comment:
This is quite confusing to read, I'd just use an if statement.

Reply:
The bind makes it all bad :smile: